rotkehlchen/chain/evm/decoding/rainbow/decoder.py

```diff
@@ -0,0 +1,388 @@
+            decoded_events: list['EvmEvent'],
+            sender: ChecksumEvmAddress,
+    ) -> tuple[Optional['EvmEvent'], Optional['EvmEvent']]:
+        """Transform the receive and spend events to rainbow events
+
+        Returns:
+            in_event (EvmEvent | None): input event, the receive the tokens
+            out_event (EvmEvent | None): output event, the spend the tokens
+        """
+        in_event = out_event = None
+        approve_event = None
+        for event in decoded_events:
+            if (event.event_type == HistoryEventType.INFORMATIONAL and event.event_subtype == HistoryEventSubType.APPROVE):  # noqa: E501
+                approve_event = event
+
+            # Find the receive event
+            if (event.event_type == HistoryEventType.RECEIVE and event.event_subtype == HistoryEventSubType.NONE and event.location_label == sender):  # noqa: E501
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.RECEIVE
+                event.counterparty = CPT_RAINBOW_SWAPS
+                event.notes = f'Receive {event.amount} {event.asset.resolve_to_asset_with_symbol().symbol} as the result of a Rainbow Swap'  # noqa: E501
+                event.address = self.router_address
+                in_event = event
+
+            # Find the spend event
+            if (event.event_type == HistoryEventType.SPEND and event.event_subtype == HistoryEventSubType.NONE and event.location_label == sender):  # noqa: E501
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.SPEND
+                event.notes = f'Swap {event.amount} {event.asset.resolve_to_asset_with_symbol().symbol} in Rainbow Swap'  # noqa: E501
+                event.counterparty = CPT_RAINBOW_SWAPS
+                event.address = self.router_address
+                out_event = event
+
+        # Remove the approve from rainbow router to aggretor
+        # to avoid show unnecessary events
+        if approve_event:
+            decoded_events.remove(approve_event)
+
+        return in_event, out_event
+
+    def _create_fee_event(
+        self,
+        tx_hash: EVMTxHash,
+        sequence_index: int,
+        timestamp: Timestamp,
+        fee_raw: int,
+        fee_asset: Asset,
+        sender: ChecksumEvmAddress,
+    ) -> 'EvmEvent':
+        """Create a new fee event for rainbow swap
+
+        Args:
+            tx_hash (EVMTxHash): transaction hash
+            sequence_index (int): event sequence index
+            timestamp (int): transaction timestamp
+            fee_raw (int): amount of fee charged in wei
+            fee_asset (Asset): fee asset
+            sender (ChecksumEvmAddress): sender address
+        """
+        fee_amount = asset_normalized_value(amount=fee_raw, asset=self._resolve_asset(fee_asset))
+        return self.base.make_event(
+            tx_hash=tx_hash,
+            sequence_index=sequence_index,
```

Comment:
instead of this check `make_event_next_index`. In general try to avoid messing with the sequence index manually since you can reach weird states easily